gce: ensureInternalInstanceGroups: reuse instance-groups for internal load balancers

abhinavdahiya opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Based on docs for internal loadbalancer here [1], backend services [2] and instances in instance-groups [3], following restrictions apply,

- Internal LB can load balance to VMs in same region, but different subnets
- Instance groups for the backend service must contain instance of the same subnet
- An instance can only belong to one load balanced instance group

It is probably useful use-case to have nodes for the cluster belong to more than one subnet. And the current setup fails to create an internal load balancer with nodes in multiple subnets.

```
I1023 22:05:24.070949       1 gce_loadbalancer_internal.go:478] ensureInternalInstanceGroup(k8s-ig--27083f8254ed83c2, us-west1-b): adding nodes: [jstuev-5hzjp-m-1.c.openshift-dev-installer.internal jstuev-5hzjp-w-b-54qkc.c.openshift-dev-installer.internal]
E1023 22:05:25.385077       1 gce_loadbalancer.go:156] Failed to EnsureLoadBalancer(jstuev-5hzjp, openshift-ingress, router-default, a79c0f796db9e4157af2e2658433b3f6, us-west1), err: googleapi: Error 400: Resource 'projects/openshift-dev-installer/zones/us-west1-b/instances/jstuev-5hzjp-w-b-54qkc' is expected to be in the subnetwork 'projects/openshift-dev-installer/regions/us-west1/subnetworks/jstubyo-master-subnet' but is in the subnetwork 'projects/openshift-dev-installer/regions/us-west1/subnetworks/jstubyo-worker-subnet'., wrongSubnetwork
```

Also the use-case that some of these nodes (machines) might be part of some internal load balancer that is not managed by k8s is also pretty valid.
for example, you might have the machines hosting the control-plane (kube-apiserver) want to be part of a separate ILB that provides access to the apiserver through LB not managed by the k8s Service type Load Balancer.
But the current setup fails to create an interal load balancer like

```
r: failed to ensure load balancer: googleapi: Error 400: INSTANCE_IN_MULTIPLE_LOAD_BALANCED_IGS - Validation failed for instance 'projects/openshift-dev-installer/zones/us-west1-a/instances/jstuev-t285j-m-0': instance may belong to at most one load-balanced instance group.
```

So the subnet limitation should be automatically handled by the k8s cloud provider, but for now allowing users to create the IGs for instances that require this special setup should definietly help, and k8s cloud provider can just use those as-is, while maintaining the membership and lifecycle for ones created by it.

This change finds pre-existing instance-groups that ONLY contain instances that belong to the cluster, uses them for the backend service. And only ensures instance-groups for remaining ones.

[1]: https://cloud.google.com/compute/docs/instance-groups/creating-groups-of-unmanaged-instances#addinstances
[2]: https://cloud.google.com/load-balancing/docs/backend-service#restrictions_and_guidance
[3]: https://cloud.google.com/compute/docs/instance-groups/creating-groups-of-unmanaged-instances#addinstances

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
GCE: Reuse pre-existing instance groups that contain cluster nodes for internal load balancers's backend
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

abhinavdahiya committed:
legacy-cloud-providers/gce/gce_fake.go: NewFakeGCECloud: make sure that the secondary zone is also part of managedZones

The DefaultTestClusterValues has two zones `ZoneName, ScondaryZoneName` set [1], but the FakeGCECloud would only use ZoneName as managedZone.

So `TestUpdateInternalLoadBalancerNodes` that adds nodes and instancegroups in both zones fails with NotFound errors

[1]: https://github.com/kubernetes/kubernetes/blob/bb052ceacb86d0603a5c2053a8c2c0c96abf83c5/staging/src/k8s.io/legacy-cloud-providers/gce/gce_fake.go#L45-L47

abhinavdahiya committed:
gce: ensureInternalInstanceGroups: reuse instance-groups for internal load balancers

Based on docs for internal loadbalancer here [1], backend services [2] and instances in instance-groups [3], following restrictions apply,

- Internal LB can load balance to VMs in same region, but different subnets
- Instance groups for the backend service must contain instance of the same subnet
- An instance can only belong to one load balanced instance group

It is probably useful use-case to have nodes for the cluster belong to more than one subnet. And the current setup fails to create an internal load balancer with nodes in multiple subnets.

```
I1023 22:05:24.070949       1 gce_loadbalancer_internal.go:478] ensureInternalInstanceGroup(k8s-ig--27083f8254ed83c2, us-west1-b): adding nodes: [jstuev-5hzjp-m-1.c.openshift-dev-installer.internal jstuev-5hzjp-w-b-54qkc.c.openshift-dev-installer.internal]
E1023 22:05:25.385077       1 gce_loadbalancer.go:156] Failed to EnsureLoadBalancer(jstuev-5hzjp, openshift-ingress, router-default, a79c0f796db9e4157af2e2658433b3f6, us-west1), err: googleapi: Error 400: Resource 'projects/openshift-dev-installer/zones/us-west1-b/instances/jstuev-5hzjp-w-b-54qkc' is expected to be in the subnetwork 'projects/openshift-dev-installer/regions/us-west1/subnetworks/jstubyo-master-subnet' but is in the subnetwork 'projects/openshift-dev-installer/regions/us-west1/subnetworks/jstubyo-worker-subnet'., wrongSubnetwork
```

Also the use-case that some of these nodes (machines) might be part of some internal load balancer that is not managed by k8s is also pretty valid.
for example, you might have the machines hosting the control-plane (kube-apiserver) want to be part of a separate ILB that provides access to the apiserver through LB not managed by the k8s Service type Load Balancer.
But the current setup fails to create an interal load balancer like

```
r: failed to ensure load balancer: googleapi: Error 400: INSTANCE_IN_MULTIPLE_LOAD_BALANCED_IGS - Validation failed for instance 'projects/openshift-dev-installer/zones/us-west1-a/instances/jstuev-t285j-m-0': instance may belong to at most one load-balanced instance group.
```

So the subnet limitation should be automatically handled by the k8s cloud provider, but for now allowing users to create the IGs for instances that require this special setup should definietly help, and k8s cloud provider can just use those as-is, while maintaining the membership and lifecycle for ones created by it.

This change finds pre-existing instance-groups that ONLY contain instances that belong to the cluster, uses them for the backend service. And only ensures instance-groups for remaining ones.

[1]: https://cloud.google.com/compute/docs/instance-groups/creating-groups-of-unmanaged-instances#addinstances
[2]: https://cloud.google.com/load-balancing/docs/backend-service#restrictions_and_guidance
[3]: https://cloud.google.com/compute/docs/instance-groups/creating-groups-of-unmanaged-instances#addinstances

smarterclayton commented:
So currently a kube cluster on GCP cannot have an externally defined LB pointing to specific nodes (for a control plane or for a node service) AND have a service LB pointing to some or all of those nodes for different services?

smarterclayton commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+			if strings.EqualFold(ig.Name, name) {
+				continue
+			}
+			instances, err := g.ListInstancesInInstanceGroup(ig.Name, zone, allInstances)
```
What's the impact of this in terms of cloud provider calls?  `O(Zones * InstanceGroupsPerZone)`, where before it was `O(Zones)`?  Will we hit novel quota / load aspects with this?

abhinavdahiya replied:
There is definitely increased API overhead from `O(zones)` -> `O(zones * instanceGroups)`

one way to reduce this would be to use certain labels/name-prefixes to request users to explicitly target discover of the instance-groups by the cluster.

but GCP doesn't support tags for instance-groups yet, https://cloud.google.com/compute/docs/labeling-resources

so our only option would be to restrict to certain prefixes similar to https://github.com/kubernetes/kubernetes/blob/ec344a1c77eea2f9320a340d97dd2de9f49dac3b/staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go#L450-L458

abhinavdahiya replied:
@smarterclayton i pushed https://github.com/kubernetes/kubernetes/pull/84466/commits/bb6e87f965820ba7e8cc2eb6657fd55f05aad026 to only turn own this external when a prefix is set to make sure we can maintain compatibility in terms of API calls for existing users and keep API calls to minimum when we want to use this new change.

smarterclayton commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+		gceZonedNodes[zone] = names.Difference(skip).UnsortedList()
+	}
+	for zone, gceNodes := range gceZonedNodes {
+		igLink, err := g.ensureInternalInstanceGroup(name, zone, gceNodes)
```
Since we already loaded instances for the instance group that matches name, can we remember it if we have it?  Or too much code complexity?

abhinavdahiya replied:
we make minimum 2 API calls per call of ensureInternalInstanceGroup
- get instance-group + create instance-group + get instance-group
- get instance-group + list instances in instance-group

keeping the logic as-such means that the not-found -> create new is encapsulated better for a fixed API overhead, otherwise we bleed the not-found to be based on ensureInternalInstanceGroups list of instance-groups

abhinavdahiya commented:
> So currently a kube cluster on GCP cannot have an externally defined LB pointing to specific nodes (for a control plane or for a node service) AND have a service LB pointing to some or all of those nodes for different services?

For a (external) TCP/UDP Network Load balancer, there is no such restriction, because it uses target-pools for backend service.

But for an Internal TCP/UDP Network Load balancer, which uses instance-groups for backend service
you cannot use SOME/SPECIFIC nodes of the cluster as backend for some non-k8s-managed Internal TCP/UDP Network Load balancer.

hopefuly that's making the restriction/distinction clearer..

abhinavdahiya committed:
gce: add ExternalInstanceGroupsPrefix to filter instance groups that will be re-used for ILB backend

Setting the prefix would allow cluster to only consider instance groups in the zone that are prefixed, reducing the number of candidates.
If the prefix is empty, no external instance groups will be used to keep backward compatibility in terms of API calls.

bowei commented:
/assign @prameshj

bowei commented:
How is this coming up? As I understand, we are already in a situation where we share the instance group among all L4 ILBs on GCP? Is there a config snippet that demonstrates the issue?

smarterclayton commented:
We run Kube clusters on top of GCP, and the control plane machines are also nodes, and they have an internal load balancer pointing to them for internal API logic.  We want to be able to use service load balancers on those control plane machines (part of work going into Kube 1.16 and 1.17 that makes that possible gain).  We found that AWS and Azure work fine for this, but GCP does not, because GCP service logic tries to put every node into an instance group, vs using instance groups that might already exist (in this case, the instance group for the control plane used by the ILB).  We'd like the GCP cloud provider to respect existing instance groups that represent nodes so that we can have both ILBs pointing to the cluster nodes and service load balancers pointing to the cluster nodes.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+			return nil, err
+		}
+		for _, ig := range igs {
+			if strings.EqualFold(ig.Name, name) {
```
shouldn't we skip this zone entirely if the name is an exact match, because the name matching means it could be the instance group created with the older style code that includes all the nodes in the zone?

abhinavdahiya replied:
> because the name matching means it could be the instance group created with the older style code that includes all the nodes in the zone

There is no guarantee that old ig has all the required nodes.
Also the ig might exist from a previous partial case, where it added some nodes..

This makes sure that we reuse as much as possible, and later on the remaining nodes will be reconciled in k8s managed IG anyways.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
 	var igLinks []string
-	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+	gceZonedNodes := map[string][]string{}
```
As i understand this change, here is what happens:

1) For each of the zones that the cluster has nodes in, list all existing instance groups.
2) if any of the instance groups contain just cluster nodes(all of them or a subset) in the zone, use that instance group. Create a new instance group with just the remaining nodes.

I see this will help in case there are control-plane nodes that are in their own instance groups. How is it helping the other use-case, where nodes are in multiple subnets?

If the use-case is just to support the control-plane scenario, can we not label those nodes with a master role? https://github.com/kubernetes/kubernetes/blob/ed5b03831385b0071b127ff6c329ae344b9780f6/pkg/controller/service/controller.go#L626

abhinavdahiya replied:
> So the subnet limitation should be automatically handled by the k8s cloud provider, but for now allowing users to create the IGs for instances that require this special setup should definietly help, and k8s cloud provider can just use those as-is, while maintaining the membership and lifecycle for ones created by it.

As from the PR description, k8s code doesn't handle nodes in different subnets, this allows users to create their IGs and k8s will reuse them until it can handle the subnets itself.

bowei replied:
Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?

abhinavdahiya replied:
> Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?

It is most definitely possible, that only applies when the nodes are master nodes, that wouldn't solve the problem when the node is actually a compute node.. ?

abhinavdahiya replied:
Also we want our master nodes to be part of the LB backend, there is no reason why the control-plane hosts cannot be part of the LB backend pool.

prameshj replied:
> > Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?
> 
> It is most definitely possible, that only applies when the nodes are master nodes, that wouldn't solve the problem when the node is actually a compute node.. ?

Is there a use-case for this? As I understood the problem, the master nodes that were part of a different LoadBalancer Instance group.

abhinavdahiya replied:
hmm...

the first case that we want to support is:
**We want masters to be part of the SLB backend.** 
But we have other services on the control-plane that cannot be backed by SLB yet, therefore we have separate externally managed LB pointing to the control-plane nodes.. and that blocks our masters need to be part of SLB as gcp only allows an instance to be part of only one instance-group that is backed by LB

reusing the instance-groups for SLB allows us to achieve external LB pointing to control-plane and SLBs pointing to control.

The second case that we want to support is:
**nodes that are backends of the SLB can be in mutiple subnets,** 
today that is not possible as the k8s controller creates and tries to put all the instances in a zone in one instance-group irrespective of the subnet.

Reusing the instance-groups allows us to manually create the instance-groups for now and let k8s create and manage the rest in the same subnet.

smarterclayton replied:
I'll note that this is unique to GCP - other cloud providers work fine when masters have SLB or external cloud LB pointed at them.  So the first part abhinav mentioned is a consistency and behavior (we wouldn't want GCP to be inconsistent with the others, etc).

On the subnets thing, there are lots of reasons for instances in a zone to be in different subnets (you may want a set of DMZ nodes targeted by internal traffic that have firewall rules to block all but a specific set of config from crossing).

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -676,6 +676,52 @@ func TestEnsureLoadBalancerDeletedSucceedsOnXPN(t *testing.T) {
+		instances[0].Instance,
+		fmt.Sprintf("projects/%s/zones/%s/instances/%s", vals.ProjectID, vals.ZoneName, "test-node-2"),
+	)
+}
```
Need to also test the case where the pre-existing ig contains non-cluster nodes, so it doesn't get picked up. 

Also the case where prefix is not set, and there is a pre-existing ig and that shouldn't get picked up.

abhinavdahiya replied:
https://github.com/kubernetes/kubernetes/pull/84466/commits/b346fe03a40178db110e2c4ef32eca23f1d0bf5c adds the unit test.

abhinavdahiya committed:
gce: skip ensureInstanceGroup for a zone that has no remaining nodes for k8s managed IG

Make sure we don't ensure cluster-managed instance group for a zone that has no more nodes left. Creating am IG for a zone that has no instances causes errors when attaching to backend service.

```
failed to ensure load balancer: googleapi: Error 400: Invalid value for field 'resource.backends[5].group': 'https://www.googleapis.com/compute/v1/projects/openshift-qe/zones/us-central1-c/instanceGroups/k8s-ig--8d8682bc12c7a717'. Instance group must have a network to be attached to a backend service. Add an instance to give the instance group a network., invalid
```

abhinavdahiya commented:
@prameshj Any updates on this??

cc @smarterclayton

prameshj commented:
I will take another look and get back on this one.
cc @freehan

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -541,9 +537,48 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+			if strings.EqualFold(ig.Name, name) {
+				continue
+			}
+			instances, err := g.ListInstancesInInstanceGroup(ig.Name, zone, allInstances)
```
Maybe move lines 562 to 575 as a helper function? Like ReuseExistingInstanceGroup(ig.Name, nodesToInclude) that returns a bool and a list of nodes in the IG?

abhinavdahiya replied:
done @prameshj PTAL https://github.com/kubernetes/kubernetes/pull/84466/commits/930cac443c139826c9e21a467fae7a6fc6a9fbdf

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -541,9 +537,48 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+			return nil, err
+		}
+		names := sets.NewString()
+		for _, h := range hosts {
```
Can we just loop through zNodes and use node.Name? Wouldn't it be the same?

abhinavdahiya replied:
So if there is an GCE Instance named `worker-0`, the internal dns name for that instance is `worker-0.<zone>.c.<project name>`
some linux OSes will use that to set the hostname to `worker-0` but some of them will keep the entire internal DNS name.
But the kubelet uses hostname to set the name of the node.

so you might receive a list of `worker-{...}.<zone>.c.<project name>`, but you want to deal with instance names ie `worker={..}` when interacting with GCE apis, therefore i'm iterating over hosts/instances.Name instead of nodes.Name

prameshj reviewed (commented):
A few nits and comments. Please rebase the test file conflicts.

abhinavdahiya commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:
```diff
@@ -129,6 +129,9 @@ type Cloud struct {
 	manager                  diskServiceManager
+
+	externalInstanceGroupsPrefix string // If non-"", finds prefixed instance groups for ILB.
+
```
I had it like that to keep it consistent with https://github.com/kubernetes/kubernetes/blob/d7e0e553277c9a7fe3d0992a0a0c511a58ca7400/staging/src/k8s.io/legacy-cloud-providers/gce/gce.go#L128

abhinavdahiya commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_instancegroup.go:
```diff
@@ -60,6 +62,21 @@ func (g *Cloud) ListInstanceGroups(zone string) ([]*compute.InstanceGroup, error
+
+	mc := newInstanceGroupMetricContext("list", zone)
+	f := filter.None
+	if prefix != "" {
```
hmm, technically currently this will return all instance groups when the prefix specified is empty, which i think should be fine as there is a logically correct response for that input. 

If you think this function should error on empty prefix I can update the comment to include that in the api?

prameshj replied:
Maybe just add a comment saying this function is called only with non-empty prefix. the behavior to list all when prefix is empty is correct, but it feels like an unnecessary step given the current usecase. Adding the comment is probably enough for now.

abhinavdahiya committed:
REVIEW: update the variable names for ensureInternalInstanceGroups and new should reuse ig function

abhinavdahiya commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:
```diff
@@ -173,6 +176,9 @@ type ConfigGlobal struct {
 	Multizone          bool     `gcfg:"multizone"`
+	// ExternalInstanceGroupsPrefix is the prefix that will be used to filter instance groups
+	// that be backend for ILB containing cluster nodes if not-empty.
+	ExternalInstanceGroupsPrefix string `gcfg:"external-instance-groups-prefix"`
```
done.

abhinavdahiya commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -541,9 +537,48 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+		for _, h := range hosts {
+			names.Insert(h.Name)
+		}
+		skip := sets.NewString()
```
done PTAL https://github.com/kubernetes/kubernetes/pull/84466/commits/930cac443c139826c9e21a467fae7a6fc6a9fbdf

abhinavdahiya commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -477,19 +477,15 @@ func (g *Cloud) ensureInternalHealthCheck(name string, svcName types.NamespacedN
-
-	gceNodes := sets.NewString()
+	gceNodes := sets.NewString(nodes...)
+	groupNodes := sets.NewString()
```
done PTAL https://github.com/kubernetes/kubernetes/pull/84466/commits/930cac443c139826c9e21a467fae7a6fc6a9fbdf

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -576,6 +603,38 @@ func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]s
+	return ret
+}
+
+// reuseInstanceGroup returns true if all the instances in the ig are part of the instancesToInclude and list of instances in the ig.
```
Fix comment "reuseInstanceGroupForInternalLoadbalancer returns".  Also include a comment about the other 2 return values.

abhinavdahiya replied:
done 3b484c45d3d

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -564,9 +560,40 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+		}
+	}
+	for zone, gceNodes := range gceZonedNodes {
+		igLink, err := g.ensureInternalInstanceGroup(name, zone, gceNodes)
```
Just to clarify - when control reaches here, gceNodes will include all nodes that are already present in 'name' IG(except an nodes that got deleted) as well as nodes to be added that were not found in other IGs.

abhinavdahiya replied:
that's correct.

abhinavdahiya commented:
@prameshj thanks for another round of review, I will try to get back on those by eow.

marun commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -785,6 +785,73 @@ func TestEnsureLoadBalancerDeletedSucceedsOnXPN(t *testing.T) {
+	require.NoError(t, err)
+
+	svc := fakeLoadbalancerService(string(LBTypeInternal))
+	_, err = gce.ensureInternalLoadBalancer(
```
How will this call to `ensureInternalLoadBalancer` succeed if svc has not been created first?

sttts commented:
Any progress here?

abhinavdahiya committed:
legacy-cloud-providers/gce: fixup function description

abhinavdahiya commented:
> https://github.com/kubernetes/kubernetes/pull/84466#discussion_r380013622
@prameshj the doc updates should be done in 3b484c45d3d

abhinavdahiya commented:
@prameshj any updates on this?

prameshj commented:
> @prameshj any updates on this?

I wasn't sure if this was ready for review. I will take another look.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -533,30 +533,15 @@ func (g *Cloud) ensureInternalHealthCheck(name string, svcName types.NamespacedN
-
-	gceNodes := sets.NewString()
+	gceNodes := sets.NewString(nodes...)
+	igNodes := sets.NewString()
```
We need to cap the total number of nodes in the instance group to 1000. That logic was added after this PR was opened, so rebase probably removed it. Can you add it back?

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -608,9 +593,40 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
+	gceZonedNodes := map[string][]string{}
 	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+		instances, err := g.getFoundInstanceByNames(nodeNames(nodes))
```
Can we split this into its own function and invoke it here instead of putting this code inline?

Something like: 
```
if g.externalInstanceGroupsPrefix != "" {
   return reuseInstanceGroups()
} 

// Keep the existing logic
}
```

This way it is clearer that the default behavior is not changed.

abhinavdahiya replied:
@prameshj pushed an update that makes it conditional. PTAL

cheftako commented:
/remove-lifecycle rotten

cheftako commented:
@abhinavdahiya are you still working on this?

abhinavdahiya committed:
gce: fixup make reuse instance groups codepath clearer

abhinavdahiya commented:
> @abhinavdahiya are you still working on this?

Yes, @cheftako I am more than happy to keep fixing this to the point we can merge this. I might be a little slow but I will try my best to keep up on this.

andrewsykim commented:
/triage accepted
/assign @cheftako

cheftako commented:
@abhinavdahiya Let us know when you are ready for another review.

abhinavdahiya commented:
> @abhinavdahiya Let us know when you are ready for another review.

@cheftako rebased and should be ready for another look. Thanks!
